### PR TITLE
Simplify target <-> string API

### PR DESCRIPTION
**Pull request by abadams**
Converting strings to targets always felt wonky to me. This is a proposed simplification of the API that just adds an explicit Target(string) constructor, and provides a separate way to validate target strings.

### CODE REVIEW

File: src/Target.cpp [hunk 1]
```diff
@@ -390,34 +333,105 @@ bool Target::merge_string(const std::string &target) {
...
+    for (auto feature_entry : feature_name_map) {
+        features += separator + feature_entry.first;
+        if (features.length() - line_char_start > 70) {
+            separator = "\n";
+            line_char_start = features.length();
+        } else {
+            separator = ", ";
+        }
+    }
+    user_error << "Did not understand HL_TARGET=" << target << "\n"
```
Comment (steven-johnson) [inline src/Target.cpp]: nit: target strings can be specified in ways other than HL_TARGET (e.g., as arguments on the command line to a generator)

File: src/Target.h [hunk 2]
```diff
@@ -81,6 +81,22 @@ struct Target {
         }
     }
 
+    /** Given a string of the form used in HL_TARGET
+     * (e.g. "x86-64-avx"), construct the Target it specifies. Note
+     * that this always starts with the result of get_host_target(),
+     * replacing only the parts found in the target string, so if you
+     * omit (say) an OS specification, the host OS will be used
+     * instead. An empty string is exactly equivalent to
+     * get_host_target().
```
Comment (steven-johnson) [inline src/Target.h]: Worth commenting that bad target strings will fail with user_error

File: src/Target.cpp [hunk 3]
```diff
@@ -390,34 +333,105 @@ bool Target::merge_string(const std::string &target) {
...
+    }
+}
+
+Target::Target(const char *s) {
+    *this = Target(std::string(s));
+}
+
+bool Target::validate_target_string(const std::string &s) {
+    Target t;
+    return merge_string(t, s);
```
Comment (steven-johnson) [inline src/Target.cpp]: It seems a little wasteful to throw away the result: if you need to check and fail in a way other than user_error, you parse twice. Not sure if this matters.

Comment (abadams) [inline src/Target.cpp]: The only thing I can find that validates target strings is the test for it, so I figured inefficiency is unimportant here. I could change it to return a pair<Target, bool>, or to take a Target * that it clobbers as the last arg?

Comment (steven-johnson) [inline src/Target.cpp]: Meh. Premature optimization. Good as is.

**Comment by steven-johnson** [comment]
LGTM

**Pull request merged**